README.md
[![Build Status](https://travis-ci.org/cubicmushroom/exceptions.svg?branch=master)](https://travis-ci.org/cubicmushroom/exceptions)

Cubic Mushroom Exceptions
=========================

This library provides just a single abstract exception class with it's own bild() method for easily creating exceptions
with various properties, passed as an array of arguments.


How to use
----------

Just have a look at the `src/CubicMushroom/Exceptions/AbstractException` class methods for basic methods, but here is a 
brief overview...

`AbstractException` provides `build()` method that can be used to prepare exceptions to be thrown.

When using `build()`, you don't need to pass in a $message, $code.  If you don't the default for the class  will be 
used.  This is returned by the classes `getDefaultMessage()` and `getDefaultCode()` methods, if they exist.
   
As it's not helpful when an exception doesn't have a message, if no message is either passed, or a default set in the 
exception class, a `MissingExceptionMessageException` will be thrown.

The `build()` method also accepts a second array parameter.  If passed, the `build()` method will attempt to set all the 
properties on the build exception using the setter for the property.  If the setter is not found, then a 
`SetterNotFoundException` will be thrown.


Roadmap
-------

1. Add way of defining required additional parameters
2. More stuff, as I think of it

CHANGELOG
# Change Log
All notable changes to this project will be documented in this file.

## [v1.0.5] - 2015-02-20
### Changed
- Reverted making AbstractException::getDefaultMessage() abstract

## [v1.0.4] - 2015-02-20
### Changed
- Made AbstractException::getDefaultMessage() abstract

## [v1.0.3] - 2015-02-02
### Changed
- Allowed AbstractException::build() method to be overridden but removing the final keyword

## [v1.0.2] - 2015-02-01
### Changed
- Require PHP 5.4 or above

## [v1.0.1] - 2015-01-31
### Added
- .gitattributes to exclude tests from archive

## [v1.0.0] - 2015-01-31
### Added
- AbstractException class
- Tests for AbstractException class
- README documentation

tests/unit/CubicMushroom/Exceptions/AbstractExceptionTest.php
<?php
namespace CubicMushroom\Exceptions;

use CubicMushroom\Exceptions\Exception\Defaults\MissingExceptionMessageException;
use CubicMushroom\Exceptions\Exception\Property\SetterNotFoundException;

class AbstractExceptionTest extends \PHPUnit_Framework_TestCase
{

    const ABSTRACT_EXCEPTION_DEFAULT_CODE = 500;


    protected function setUp()
    {
    }


    protected function tearDown()
    {
    }


    // -----------------------------------------------------------------------------------------------------------------
    // Without Constructor Defaults Tests
    // -----------------------------------------------------------------------------------------------------------------

    /**
     * Tests that a class with no defaults when no arguments passed
     */
    public function testBuildWithNoDefaultsAndNoParameters()
    {
        try {
            ExtendedWithoutDefaultsException::build();
        } catch (MissingExceptionMessageException $thrownException) {
        }

        /** @noinspection PhpUndefinedVariableInspection */
        $this->assertNotEmpty($thrownException);
        $this->assertInstanceOf(get_class(new MissingExceptionMessageException), $thrownException);
        $this->assertEquals(MissingExceptionMessageException::DEFAULT_MESSAGE, $thrownException->getMessage());
    }


    /**
     * Tests that a class with no defaults when only a message is passed
     */
    public function testBuildWithNoDefaultsOnlyMessage()
    {
        $message = 'testBuildWithNoDefaultsOnlyMessage message';

        $builtException = ExtendedWithoutDefaultsException::build(array('message' => $message));

        $this->assertNotEmpty($builtException);
        $this->assertInstanceOf(get_class(new ExtendedWithoutDefaultsException), $builtException);
        $this->assertEquals($message, $builtException->getMessage());
        $this->assertEquals(self::ABSTRACT_EXCEPTION_DEFAULT_CODE, $builtException->getCode());
        $this->assertEmpty($builtException->getPrevious());
    }


    /**
     * Tests that a class with no defaults when only a code is passed
     */
    public function testBuildWithNoDefaultsOnlyCode()
    {
        $code = 1000;

        try {
            ExtendedWithoutDefaultsException::build(array('code' => $code));
        } catch (MissingExceptionMessageException $thrownException) {
        }

        /** @noinspection PhpUndefinedVariableInspection */
        $this->assertNotEmpty($thrownException);
        $this->assertInstanceOf(get_class(new MissingExceptionMessageException), $thrownException);
        $this->assertEquals(MissingExceptionMessageException::DEFAULT_MESSAGE, $thrownException->getMessage());
    }


    /**
     * Tests that a class with no defaults when only a previous exception is passed
     */
    public function testBuildWithNoDefaultsOnlyPrevious()
    {
        $previous = new PreviousException;

        try {
            ExtendedWithoutDefaultsException::build(array('previous' => $previous));
        } catch (MissingExceptionMessageException $thrownException) {
        }

        /** @noinspection PhpUndefinedVariableInspection */
        $this->assertNotEmpty($thrownException);
        $this->assertInstanceOf(get_class(new MissingExceptionMessageException), $thrownException);
        $this->assertEquals(MissingExceptionMessageException::DEFAULT_MESSAGE, $thrownException->getMessage());
    }


    /**
     * Tests that a class with no defaults when only a message and code are passed
     */
    public function testBuildWithNoDefaultsOnlyMessageAndCode()
    {
        $message = 'testBuildWithNoDefaultsOnlyMessage message';
        $code = 1000;

        $builtException = ExtendedWithoutDefaultsException::build(array('message' => $message, 'code' => $code));

        $this->assertNotEmpty($builtException);
        $this->assertInstanceOf(get_class(new ExtendedWithoutDefaultsException), $builtException);
        $this->assertEquals($message, $builtException->getMessage());
        $this->assertEquals($code, $builtException->getCode());
        $this->assertEmpty($builtException->getPrevious());
    }


    /**
     * Tests that a class with no defaults when only a message and previous exception are passed
     */
    public function testBuildWithNoDefaultsOnlyMessageAndPrevious()
    {
        $message = 'testBuildWithNoDefaultsOnlyMessage message';
        $previous = new PreviousException;

        $builtException = ExtendedWithoutDefaultsException::build(array(
            'message'  => $message,
            'previous' => $previous
        ));

        $this->assertNotEmpty($builtException);
        $this->assertInstanceOf(get_class(new ExtendedWithoutDefaultsException), $builtException);
        $this->assertEquals($message, $builtException->getMessage());
        $this->assertEquals($previous, $builtException->getPrevious());
        $this->assertEquals(self::ABSTRACT_EXCEPTION_DEFAULT_CODE, $builtException->getCode());
    }


    /**
     * Tests that a class with no defaults when only a code and previous exception are passed
     */
    public function testBuildWithNoDefaultsOnlyCodeAndPrevious()
    {
        $previous = new PreviousException;
        $code = 1000;

        try {
            ExtendedWithoutDefaultsException::build(array('code' => $code, 'previous' => $previous));
        } catch (MissingExceptionMessageException $thrownException) {
        }

        /** @noinspection PhpUndefinedVariableInspection */
        $this->assertNotEmpty($thrownException);
        $this->assertInstanceOf(get_class(new MissingExceptionMessageException), $thrownException);
        $this->assertEquals(MissingExceptionMessageException::DEFAULT_MESSAGE, $thrownException->getMessage());
        $this->assertEquals(self::ABSTRACT_EXCEPTION_DEFAULT_CODE, $thrownException->getCode());
        $this->assertEmpty($thrownException->getPrevious());
    }


    /**
     * Tests that a class with no defaults when only all exception constructor arguments are passed
     */
    public function testBuildWithNoDefaultsAllConstructorArguments()
    {
        $message = 'testBuildWithNoDefaultsOnlyMessage message';
        $code = 1000;
        $previous = new PreviousException;

        $builtException = ExtendedWithoutDefaultsException::build(array(
            'message'  => $message,
            'code'     => $code,
            'previous' => $previous,
        ));

        $this->assertNotEmpty($builtException);
        $this->assertInstanceOf(get_class(new ExtendedWithoutDefaultsException), $builtException);
        $this->assertEquals($message, $builtException->getMessage());
        $this->assertEquals($code, $builtException->getCode());
        $this->assertEquals($previous, $builtException->getPrevious());
    }


    // -----------------------------------------------------------------------------------------------------------------
    // With Constructor Defaults Tests
    // -----------------------------------------------------------------------------------------------------------------

    /**
     * Tests the build of an exception that has defaults with only a message
     */
    public function testBuildWithDefaultMessageOnlyMessage()
    {
        $message = 'testBuildWithDefaultMessageOnlyMessage message';

        $builtException = ExtendedWithDefaultsException::build(array('message' => $message));

        $this->assertNotEmpty($builtException);
        $this->assertEquals($message, $builtException->getMessage());
        $this->assertEquals(ExtendedWithDefaultsException::DEFAULT_CODE, $builtException->getCode());
        $this->assertEmpty($builtException->getPrevious());
    }


    /**
     * Tests the build of an exception that has defaults with only a code
     */
    public function testBuildWithDefaultMessageOnlyCode()
    {
        $code = 1000;

        $builtException = ExtendedWithDefaultsException::build(array('code' => $code));

        $this->assertNotEmpty($builtException);
        $this->assertEquals($code, $builtException->getCode());
        $this->assertEquals(ExtendedWithDefaultsException::DEFAULT_MESSAGE, $builtException->getMessage());
        $this->assertEmpty($builtException->getPrevious());
    }


    // Tests the build of an exception that has defaults with only a previous exception
    public function testBuildWithDefaultMessageOnlyPrevious()
    {
        $previous = new PreviousException;

        $builtException = ExtendedWithDefaultsException::build(array('previous' => $previous));

        $this->assertNotEmpty($builtException);
        $this->assertEquals($previous, $builtException->getPrevious());
        $this->assertEquals(ExtendedWithDefaultsException::DEFAULT_MESSAGE, $builtException->getMessage());
        $this->assertEquals(ExtendedWithDefaultsException::DEFAULT_CODE, $builtException->getCode());
    }


    /**
     * Tests the build of an exception that has defaults with only a message and code
     */
    public function testBuildWithDefaultMessageOnlyMessageAndCode()
    {
        $message = 'testBuildWithDefaultMessageOnlyMessageAndCode message';
        $code = 1000;

        $builtException = ExtendedWithDefaultsException::build(array('message' => $message, 'code' => $code));

        $this->assertNotEmpty($builtException);
        $this->assertEquals($message, $builtException->getMessage());
        $this->assertEquals($code, $builtException->getCode());
        $this->assertEmpty($builtException->getPrevious());
    }


    /**
     * Tests the build of an exception that has defaults with only a message and previous exception
     */
    public function testBuildWithDefaultMessageOnlyMessageAndPrevious()
    {
        $message = 'testBuildWithDefaultMessageOnlyMessageAndPrevious message';
        $previous = new PreviousException;

        $builtException = ExtendedWithDefaultsException::build(array('message' => $message, 'previous' => $previous));

        $this->assertNotEmpty($builtException);
        $this->assertEquals($message, $builtException->getMessage());
        $this->assertEquals($previous, $builtException->getPrevious());
        $this->assertEquals(ExtendedWithDefaultsException::DEFAULT_CODE, $builtException->getCode());
    }


    /**
     * Tests the build of an exception that has defaults with only a code and previous exception
     */
    public function testBuildWithDefaultMessageOnlyCodeAndPrevious()
    {
        $code = 1000;
        $previous = new PreviousException;

        $builtException = ExtendedWithDefaultsException::build(array('code' => $code, 'previous' => $previous));

        $this->assertNotEmpty($builtException);
        $this->assertEquals($code, $builtException->getCode());
        $this->assertEquals($previous, $builtException->getPrevious());
        $this->assertEquals(ExtendedWithDefaultsException::DEFAULT_MESSAGE, $builtException->getMessage());
    }


    // Tests the build of an exception that has defaults with all constructor arguments
    public function testBuildWithDefaultMessageAllConstructorArguments()
    {
        $message = 'testBuildWithDefaultMessageAllConstructorArguments message';
        $code = 1000;
        $previous = new PreviousException;

        $builtException = ExtendedWithDefaultsException::build(array(
            'message'  => $message,
            'code'     => $code,
            'previous' => $previous
        ));

        $this->assertNotEmpty($builtException);
        $this->assertEquals($message, $builtException->getMessage());
        $this->assertEquals($code, $builtException->getCode());
        $this->assertEquals($previous, $builtException->getPrevious());
    }


    // -----------------------------------------------------------------------------------------------------------------
    // Additional Arguments Tests
    // -----------------------------------------------------------------------------------------------------------------

    /**
     * Test valid additional arguments
     */
    public function testValidAdditionalArguments()
    {
        $additionalArgs = [
            'plain' => 'Single word property',
            'camelCase' => 'A camel cased property',
            'publicWithSetter' => 'A public property with a setter',
            // Not including public properties without setters at the moment.  May add these later.
            //'publicWithoutSetter' => 'A public property without a setter',
        ];

        $builtException = ExtendedAdditionalPropertiesException::build([], $additionalArgs);

        $this->assertNotEmpty($builtException);
        $this->assertEquals($additionalArgs['plain'], $builtException->getPlain());
        $this->assertEquals($additionalArgs['camelCase'], $builtException->getCamelCase());
        $this->assertEquals($additionalArgs['publicWithSetter'], $builtException->getPublicWithSetter());
        $this->assertEquals(1, $builtException->setterCalls['plain']);
        $this->assertEquals(1, $builtException->setterCalls['camelCase']);
        $this->assertEquals(1, $builtException->setterCalls['publicWithSetter']);
    }


    /**
     * Test invalid additional arguments
     *
     * @expectedException \CubicMushroom\Exceptions\Exception\Property\SetterNotFoundException
     * @expectedExceptionCode 500
     * @expectedExceptionMessage Setter for property for CubicMushroom\Exceptions\ExtendedAdditionalPropertiesException::doesNotExist not found
     */
    public function testInvalidAdditionalArguments()
    {
        $additionalArgs = [
            'doesNotExist' => 'This property does not exist',
        ];

        ExtendedAdditionalPropertiesException::build([], $additionalArgs);
    }
}


// ---
// Test classes
// ---

/**
 * Class PreviousException
 *
 * Unique classed to be used as the previous exception to avoid possible false positives when checking classes
 *
 * @package CubicMushroom\Exceptions
 */
class PreviousException extends \Exception
{
}


/**
 * Class ExtendedWithoutDefaultsException
 *
 * Exception class that does not include a default message or code getter
 *
 * @package CubicMushroom\Exceptions
 */
class ExtendedWithoutDefaultsException extends AbstractException
{
}


/**
 * Class ExtendedWithoutDefaultsException
 *
 * Exception class that includes default message and code getter
 *
 * @package CubicMushroom\Exceptions
 */
class ExtendedWithDefaultsException extends AbstractException
{

    const DEFAULT_MESSAGE = 'ExtendedWithDefaultsException message';

    const DEFAULT_CODE = 1100;


    /**
     * @param array $additionalProperties Additional properties passed to the build() method
     *
     * @return string
     */
    protected static function getDefaultMessage(array $additionalProperties)
    {
        return self::DEFAULT_MESSAGE;
    }


    protected static function getDefaultCode()
    {
        return self::DEFAULT_CODE;
    }
}


/**
 * Class ExtendedWithoutDefaultsException
 *
 * Exception class that includes default message and code getter
 *
 * @package CubicMushroom\Exceptions
 */
class ExtendedAdditionalPropertiesException extends AbstractException
{

    public $publicWithSetter;

    public $publicWithoutSetter;

    protected $plain;

    protected $camelCase;

    public $setterCalls = [
        'plain'            => 0,
        'camelCase'        => 0,
        'publicWithSetter' => 0
    ];


    /**
     * @return mixed
     */
    public function getPlain()
    {
        return $this->plain;
    }


    /**
     * @param mixed $plain
     */
    public function setPlain($plain)
    {
        $this->setterCalls['plain']++;
        $this->plain = $plain;
    }


    /**
     * @return mixed
     */
    public function getCamelCase()
    {
        return $this->camelCase;
    }


    /**
     * @param mixed $camelCase
     */
    public function setCamelCase($camelCase)
    {
        $this->setterCalls['camelCase']++;
        $this->camelCase = $camelCase;
    }


    /**
     * @param array $additionalProperties Additional properties passed to the build() method
     *
     * @return string
     */
    protected static function getDefaultMessage(array $additionalProperties)
    {
        return 'ExtendedAdditionalPropertiesException message';
    }


    protected static function getDefaultCode()
    {
        return 1200;
    }


    /**
     * @return mixed
     */
    public function getPublicWithSetter()
    {
        return $this->publicWithSetter;
    }


    /**
     * @param mixed $publicWithSetter
     */
    public function setPublicWithSetter($publicWithSetter)
    {
        $this->setterCalls['publicWithSetter']++;
        $this->publicWithSetter = $publicWithSetter;
    }
}

gulpfile.js
var gulp        = require('gulp'),
    bump        = require('gulp-bump'),
    codecept    = require('gulp-codeception'),
    confirm     = require('gulp-confirm'),
    filter      = require('gulp-filter'),
    git         = require('gulp-git'),
    notify      = require('gulp-notify'),
    shell       = require('gulp-shell'),
    tag_version = require('gulp-tag-version'),
    watch       = require('gulp-watch');

var srcFilePattern              = './src/**/*.php',
    unitTestPattern             = './tests/unit/**/*.php',
    testSupportFilePattern      = './tests/_support/**/*.php',
    changelogSupportFilePattern = './CHANGELOG',
    versionFilePattern          = ['./composer.json', './package.json'];


gulp.task('clear', shell.task(
    [
        'echo', 'echo', 'echo', 'echo', 'echo', 'echo', 'echo', 'echo', 'echo', 'echo', 'echo', 'echo', 'echo',
        'echo', 'echo', 'echo', 'echo', 'echo', 'echo', 'echo', 'echo', 'echo', 'echo', 'echo', 'echo', 'echo',
        'echo', 'echo', 'echo', 'echo', 'echo', 'echo', 'echo', 'echo', 'echo', 'echo', 'echo', 'echo', 'echo'
    ]
));


/**
 * Codeception tasks
 */
gulp.task('cc:unit', function () {
    var options = {
        //flags    : '--no-colors',
        testSuite: 'unit',
        debug    : false,
        notify   : true
    };
    gulp.src(unitTestPattern)
        .pipe(codecept(false, options))
        .on('error', notify.onError({
            title  : "Unit Tests failed!",
            message: "Errors during runtime <%= error.message %>",
            icon   : './node_modules/gulp-codeception/assets/test-fail.jpg'
        }))
        .pipe(notify({
            title  : 'Success!',
            icon   : './node_modules/gulp-codeception/assets/test-pass.jpg',
            message: 'Everything was successful!'
        }))
});
gulp.task('watch:cc:unit', ['cc:unit'], function () {
    gulp.watch([srcFilePattern, unitTestPattern, testSupportFilePattern], ['clear', 'cc:unit']);
});


/**
 * Version Tagging tasks
 */
/**
 * Bumping version number and tagging the repository with it.
 * Please read http://semver.org/
 *
 * You can use the commands
 *
 *     gulp patch     # makes v0.1.0 → v0.1.1
 *     gulp feature   # makes v0.1.1 → v0.2.0
 *     gulp release   # makes v0.2.1 → v1.0.0
 *
 * To bump the version numbers accordingly after you did a patch,
 * introduced a feature or made a backwards-incompatible release.
 */

function incrementVersion(importance) {
    // get all the files to bump version in
    return gulp.src(versionFilePattern)
        // bump the version number in those files
        .pipe(bump({type: importance}))
        // save it back to filesystem
        .pipe(gulp.dest('./'))
        // commit the changed version number
        .pipe(git.commit('Released updated version'))
        // read only one file to get the version number
        .pipe(filter('package.json'))
        // **tag it in the repository**
        .pipe(tag_version());
}

gulp.task('confirm:changelog', function () {
    return gulp.src(versionFilePattern)
        .pipe(confirm({
            question  : 'Have you updated the CHANGELOG?',
            "continue": function (answer) {
                return answer.toLowerCase() === 'y';
            }
        }));
});
gulp.task('commit:changelog', function () {
    return gulp.src(changelogSupportFilePattern)
        .pipe(git.commit('Updating CHANGELOG'));
});
gulp.task('version:patch', ['confirm:changelog'], function () {
    return incrementVersion('patch');
});
gulp.task('version:feature', ['confirm:changelog'], function () {
    return incrementVersion('minor');
});
gulp.task('version:release', ['confirm:changelog'], function () {
    return incrementVersion('major');
});

src/CubicMushroom/Exceptions/Exception/Property/SetterNotFoundException.php
<?php
/**
 * Created by PhpStorm.
 * User: toby
 * Date: 31/01/15
 * Time: 15:38
 */

namespace CubicMushroom\Exceptions\Exception\Property;

use CubicMushroom\Exceptions\AbstractException;

class SetterNotFoundException extends AbstractException
{

    /**
     * @var string
     */
    protected $class;

    /**
     * Name of the property for which the setter is missing
     *
     * @var string
     */
    protected $property;


    /**
     * @param array $additionalProperties Additional properties passed to the build() method
     *
     * @return string
     */
    protected static function getDefaultMessage(array $additionalProperties)
    {
        return "Setter for property for {$additionalProperties['class']}::{$additionalProperties['property']} not " .
               "found";
    }


    /**
     * @return int
     */
    protected static function getDefaultCode()
    {
        return 500;
    }


    // -----------------------------------------------------------------------------------------------------------------
    // Getters and Setters
    // -----------------------------------------------------------------------------------------------------------------

    /**
     * @return string
     */
    public function getClass()
    {
        return $this->class;
    }


    /**
     * @param string $class
     */
    public function setClass($class)
    {
        $this->class = $class;
    }


    /**
     * @return string
     */
    public function getProperty()
    {
        return $this->property;
    }


    /**
     * @param string $property
     */
    public function setProperty($property)
    {
        $this->property = $property;
    }
}

src/CubicMushroom/Exceptions/AbstractException.php
<?php
/**
 * Created by PhpStorm.
 * User: toby
 * Date: 31/01/15
 * Time: 11:36
 */

namespace CubicMushroom\Exceptions;

use CubicMushroom\Exceptions\Exception\Defaults\MissingExceptionMessageException;
use CubicMushroom\Exceptions\Exception\Property\SetterNotFoundException;

/**
 * Class AbstractException
 *
 * Basic abstract class that provieds a handy static build() class that allows you to
 *
 * @package CubicMushroom\Exceptions
 */
class AbstractException extends \Exception
{
    /**
     * Helper method to prepare all but the most complex exceptions
     *
     * Pass in an array containing any or all of the following keys
     * - message (string)
     * - code (int)
     * - previous (\Exception)
     *
     * If you don't pass in a 'message' or a 'code' then the method will attempt to get these from the exception class
     * getDefaultMessage or getDefaultCode methods, respectively.
     *
     * If no message is prepared, then we throw a \RuntimeException... There's no excuse for not including a useful
     * message!
     *
     * If no code if prepared, then 500 is used
     *
     * @param array $constructorArguments [optional] Arguments to be passed to the constructor
     * @param array $additionalProperties [optional] Additional p
     *
     * @return static
     *
     * @throws \RuntimeException if no message can be prepared... There's no excuse for not including a useful message!
     *
     */
    public static function build(array $constructorArguments = array(), $additionalProperties = array())
    {
        $constructorDefaults = array('message' => null, 'code' => null, 'previous' => null);
        $constructorArguments = array_merge($constructorDefaults, $constructorArguments);

        $message = self::prepareMessage($constructorArguments['message'], $additionalProperties);
        $code = self::prepareCode($constructorArguments['code']);
        $previous = $constructorArguments['previous'];

        $e = new static($message, $code, $previous);

        $e = self::addAdditionalProperties($e, $additionalProperties);

        return $e;
    }


    /**
     * Prepares the message from either its passed value, or the class' default message, accessed via it's
     * getDefaultMessage(), if it exists
     *
     * @param string $message              Message passed in to build()
     * @param array  $additionalProperties Additional properties passed to the build() method
     *
     * @return string
     * @throws MissingExceptionMessageException
     */
    protected static function prepareMessage($message, $additionalProperties)
    {
        if (empty($message)) {
            $message = static::getDefaultMessage($additionalProperties);
        }

        return $message;
    }


    /**
     * @param array $additionalProperties Additional properties passed to the build() method
     *
     * @return string Should return a string on child classes.  This class throws an exception, as there is no default
     *                message for the abstract class
     *
     * @throws MissingExceptionMessageException
     */
    protected static function getDefaultMessage(
        /** @noinspection PhpUnusedParameterInspection */
        array $additionalProperties
    ) {
        throw MissingExceptionMessageException::build();
    }


    /**
     * @param int $code Code passed in to build()
     *
     * @return int
     */
    protected static function prepareCode($code)
    {
        if (empty($code)) {
            $code = static::getDefaultCode();
        }

        return $code;
    }


    /**
     * @return int
     */
    protected static function getDefaultCode()
    {
        return 500;
    }


    /**
     * @param AbstractException $e    Newly created exception
     * @param array             $args Arguments passed to build to be added (see noted on build() method)
     *
     * @return static
     *
     * @throws SetterNotFoundException of
     */
    protected static function addAdditionalProperties(AbstractException $e, $args)
    {
        foreach ($args as $property => $value) {
            $setter = 'set' . ucfirst($property);
            if (!is_callable([$e, $setter])) {
                throw SetterNotFoundException::build(array(), array('class' => get_class($e), 'property' => $property));
            }

            $e->$setter($value);
        }

        return $e;
    }


    /**
     * Returns the Exception that is the root cause of one or more subsequent exceptions.
     *
     * @param \Exception $exception The exception of which to find a base exception.
     *
     * @return \Exception
     *
     * @see http://bit.ly/16bdp00 <= This is where I found this method
     */
    final public static function getBaseException(\Exception $exception)
    {
        while ($exception->getPrevious() !== null) {
            $exception = $exception->getPrevious();
        }

        return $exception;
    }
}